lib/time/time.go

```diff
@@ -0,0 +1,548 @@
+package time
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"go.starlark.net/starlark"
+	"go.starlark.net/starlarkstruct"
+	"go.starlark.net/syntax"
+)
+
+// ModuleName defines the expected name for this Module when used in the
+// starlark runtime.
+const ModuleName = "time"
+
+// Module time is a Starlark module of time-related functions.
+var Module = &starlarkstruct.Module{
+	Name: ModuleName,
+	Members: starlark.StringDict{
+		"duration":       starlark.NewBuiltin("duration", newDuration),
+		"parse_duration": starlark.NewBuiltin("parse_duration", parseDuration),
+		"parse_location": starlark.NewBuiltin("parse_location", parseLocation),
+		"now":            starlark.NewBuiltin("now", now),
+		"struct":         starlark.NewBuiltin("struct", starlarkstruct.Make),
+		"time":           starlark.NewBuiltin("time", newTime),
+		"parse_time":     starlark.NewBuiltin("time", parseTime),
+		"sleep":          starlark.NewBuiltin("sleep", sleep),
+		"from_timestamp": starlark.NewBuiltin("fromtimestamp", fromUnixTime),
+
+		"zero": Time{},
+
+		"nanosecond":  Duration(time.Nanosecond),
+		"microsecond": Duration(time.Microsecond),
+		"millisecond": Duration(time.Millisecond),
+		"second":      Duration(time.Second),
+		"minute":      Duration(time.Minute),
+		"hour":        Duration(time.Hour),
+	},
+}
+
+// LoadModule loads the time module.
+// It is concurrency-safe and idempotent.
+func LoadModule() (starlark.StringDict, error) {
+	return starlark.StringDict{
+		ModuleName: Module,
+	}, nil
+}
+
+// NowFunc is a function that generates the current time. Intentionally exported
+// so that it can be overridden, for example by applications that require their
+// Starlark scripts to be fully deterministic.
+var NowFunc = time.Now
+
+func parseDuration(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var d Duration
+	err := starlark.UnpackArgs("duration", args, kwargs, "d", &d)
+	return d, err
+}
+
+func parseLocation(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var s string
+	if err := starlark.UnpackArgs("location", args, kwargs, "s", &s); err != nil {
+		return nil, err
+	}
+	loc, err := time.LoadLocation(s)
+	if err != nil {
+		return nil, err
+	}
+
+	return starlark.String(loc.String()), nil
+}
+
+// SleepFunc is a function that pauses the current goroutine for at least d.
+// Intentionally exported so that it can be overridden, for example by
+// applications that require their Starlark scripts to be fully deterministic.
+var SleepFunc = time.Sleep
+
+func sleep(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var dur Duration
+	if err := starlark.UnpackPositionalArgs("sleep", args, kwargs, 1, &dur); err != nil {
+		return starlark.None, err
+	}
+
+	SleepFunc(time.Duration(dur))
```

Comment:
You need to check `SleepFunc` being not nil otherwise we might segfault here.

Reply:
Personally I think that if users set `SleepFunc` to nil and try to call `SleepFunc` the program should panic.

Reply:
Well I don't like panicking on this level. Imagine you expose setting the sleep function to the user script (for whatever reason). Now your whole program will explode without any way for the programmer to intervene. IMO we should either throw an error here or use the default sleep function in presence of `SleepFunc` being `nil`.

Reply:
I agree with @b5 that it's fine to panic here, as it's an application programming error.

Reply:
Ok, but this means we never should expose setting `SleepFunc` to the interpreter and thus to the user. To be honest I would make `SleepFunc == nil`  the default and then run `time.Sleep` to circumvent all of this, but it's not my decision. :-)

Reply:
It's fine to insist that the value be non-nil, and simply panic if someone violates that. Users: don't do that.